CONTRIBUTING.md

```diff
@@ -3,11 +3,203 @@ Guidelines for Contributing
+- Use our [Bugzilla bug tracker](http://d.puremagic.com/issues/)
+- Follow the [D style](http://dlang.org/dstyle.html)
+- Participate in [our forum](http://forum.dlang.org/)
+- Ask questions on `#d` IRC channel on freenode.org ([web interface](https://kiwiirc.com/client/irc.freenode.net/d))
+- Review Phobos additions in the [Review Queue](http://wiki.dlang.org/Review_Queue)
+
+How to make a good submission?
+------------------------------
+
+The usual workflow is:
+
+1. [Fork](https://help.github.com/articles/fork-a-repo/) [Phobos](https://github.com/D-Programming-Language/phobos) on Github
+2. Create your [own branch](https://github.com/Kunena/Kunena-Forum/wiki/Create-a-new-branch-with-git-and-manage-branches)
+3. [Work locally](http://wiki.dlang.org/Starting_as_a_Contributor#Building_D) on your new feature or fix (see the tips below)
+4. [Test](#tests) your improvements locally
+5. Submit your [pull request](https://help.github.com/articles/creating-a-pull-request/) (PR)
+
+Below is a list of tips and checks that you should consider before opening your pull request:
+
+### Code style
+
+- [ ] Did you stick to the [style guide](http://dlang.org/dstyle.html)? (e.g. braces in their own lines (allman style) or white spaces between operators)
+- Avoid unnecessarily importing other Phobos modules
+- Try to not require a garbage collector (this is not a must)
+- Avoid code duplication
+- Maximal `@nogc`, `@safe`, `pure`, `nothrow` in _non-templated_ functions
+- Don't add `@nogc`, `@safe`, `pure`, `nothrow` attributes to _templated_ functions - let the compiler infer it! However add unittest checks for these attributes if possible (e.g. `pure nothrow @nogc @safe unittest { ... }`)
+- Avoid to use `auto` as return type wherever possible (makes the documentation easier to read)
+- Use `static` nested structs for ranges (aka [Voldemort types](http://wiki.dlang.org/Voldemort_types)) and thus store function parameters explicitly in the nested struct
+
+### Code advice
+
+- Don't be afraid to use the new compiler or language features (e.g. `Foo!Range` vs. `Foo!(Range)`)
+- Be critical when looking at Phobos code (some parts of Phobos are a bit older and have some relics due to lacking functionality of the compiler)
+- [ ] Is there an easier or more performant way to do it?
+- [ ] Does your PR address only one topic? Can it be split up in smaller PRs?
+- [ ] Is your code flexible enough to accommodate more use cases?
+- [ ] Read through your code again - is there any part that is not understandable?
+
+### Tests
+
+- [Autotester](http://wiki.dlang.org/Git_Commit_Tester) will automatically compile the code in your PR and test it on all supported platforms. For your first PR you need approval from any reviewer - ping them in case they forget.
+- [ ] Do all [tests pass locally](http://wiki.dlang.org/Starting_as_a_Contributor#Unittest_phobos)? You can run the tests of
+	- single module with `rdmd -main -unittest module.d`
```

Comment:
shouldn't testing a single module use `make -f posix.mak std/module.test` as well? Using `rdmd` you might be linking to your system's phobos library rather than your local development version.